File: mobile-app/lib/models/work_start.dart
/// The two daily activities that have a start deadline.
enum WorkStartType {
  chane('chane', 'شروع چانه‌گیری'),
  baking('baking', 'شروع پخت');

  const WorkStartType(this.apiValue, this.label);

  final String apiValue;
  final String label;

  static WorkStartType fromApi(String? value) => WorkStartType.values.firstWhere(
        (t) => t.apiValue == value,
        orElse: () => WorkStartType.chane,
      );
}

/// One activity's state for today: whether it has been ticked, whether it
/// was late, and how long is left before the deadline.
class WorkStartItem {
  const WorkStartItem({
    required this.type,
    required this.label,
    required this.deadline,
    required this.started,
    required this.isLate,
    required this.lateMinutes,
    required this.overdue,
    required this.isHoliday,
    this.startedAt,
    this.startedBy,
    this.warning,
    this.minutesRemaining,
  });

  final WorkStartType type;
  final String label;
  final String deadline;
  final bool started;
  final bool isLate;
  final int lateMinutes;
  final bool overdue;
  final bool isHoliday;
  final String? startedAt;
  final String? startedBy;
  final String? warning;
  final int? minutesRemaining;

  /// True when the deadline is close enough to be worth warning about.
  bool get isApproaching =>
      !started &&
      !isHoliday &&
      minutesRemaining != null &&
      minutesRemaining! >= 0 &&
      minutesRemaining! <= 20;

  factory WorkStartItem.fromJson(Map<String, dynamic> json) {
    return WorkStartItem(
      type: WorkStartType.fromApi(json['type'] as String?),
      label: json['label'] as String? ?? '',
      deadline: json['deadline'] as String? ?? '',
      started: json['started'] == true,
      isLate: json['is_late'] == true,
      lateMinutes: _int(json['late_minutes']),
      overdue: json['overdue'] == true,
      isHoliday: json['is_holiday'] == true,
      startedAt: json['started_at'] as String?,
      startedBy: json['started_by'] as String?,
      warning: json['warning'] as String?,
      minutesRemaining: json['minutes_remaining'] == null
          ? null
          : _int(json['minutes_remaining']),
    );
  }

  static int _int(dynamic value) =>
      value is num ? value.toInt() : int.tryParse('$value') ?? 0;
}

/// The published late-start rules, shown to every member of staff so the
/// tariff is known in advance rather than discovered on payday.
class LateTariff {
  const LateTariff({
    required this.freeDays,
    required this.summary,
    required this.tier1Formatted,
    required this.tier2Formatted,
  });

  final int freeDays;
  final String summary;
  final String tier1Formatted;
  final String tier2Formatted;

  factory LateTariff.fromJson(Map<String, dynamic> json) {
    return LateTariff(
      freeDays: WorkStartItem._int(json['free_days']),
      summary: json['summary'] as String? ?? '',
      tier1Formatted: json['tier1_amount_formatted'] as String? ?? '',
      tier2Formatted: json['tier2_amount_formatted'] as String? ?? '',
    );
  }
}

/// This month's late days and what they have cost so far.
class LateMonthSummary {
  const LateMonthSummary({
    required this.periodLabel,
    required this.lateDays,
    required this.warningsRemaining,
    required this.penaltyFormatted,
    required this.nextDayAmountFormatted,
  });

  final String periodLabel;
  final int lateDays;
  final int warningsRemaining;
  final String penaltyFormatted;
  final String nextDayAmountFormatted;

  factory LateMonthSummary.fromJson(Map<String, dynamic> json) {
    return LateMonthSummary(
      periodLabel: json['period_label'] as String? ?? '',
      lateDays: WorkStartItem._int(json['late_days']),
      warningsRemaining: WorkStartItem._int(json['warnings_remaining']),
      penaltyFormatted: json['penalty_total_formatted'] as String? ?? '',
      nextDayAmountFormatted:
          json['next_day_amount_formatted'] as String? ?? '',
    );
  }
}

/// Today's board for both activities.
class WorkStartBoard {
  const WorkStartBoard({
    required this.dateDisplay,
    required this.isHoliday,
    required this.items,
    this.tariff,
    this.monthSummary,
  });

  final String dateDisplay;
  final bool isHoliday;
  final List<WorkStartItem> items;
  final LateTariff? tariff;
  final LateMonthSummary? monthSummary;

  WorkStartItem? of(WorkStartType type) {
    for (final item in items) {
      if (item.type == type) return item;
    }
    return null;
  }

  /// Anything late or already past its deadline, for the warning banner.
  List<WorkStartItem> get problems =>
      items.where((i) => i.isLate || i.overdue).toList();

  factory WorkStartBoard.fromJson(Map<String, dynamic> json) {
    return WorkStartBoard(
      dateDisplay: json['date_display'] as String? ?? '',
      isHoliday: json['is_holiday'] == true,
      items: ((json['items'] as List?) ?? const [])
          .cast<Map<String, dynamic>>()
          .map(WorkStartItem.fromJson)
          .toList(),
      tariff: json['tariff'] is Map
          ? LateTariff.fromJson(
              (json['tariff'] as Map).cast<String, dynamic>())
          : null,
      monthSummary: json['month_summary'] is Map
          ? LateMonthSummary.fromJson(
              (json['month_summary'] as Map).cast<String, dynamic>())
          : null,
    );
  }
}

File: backend/app/Models/WorkStart.php
<?php

namespace App\Models;

use App\Models\Concerns\BelongsToBakery;
use App\Support\AppCalendar;
use App\Support\CurrentBakery;
use App\Support\Jalali;
use App\Support\LatePenalty;
use App\Support\Money;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Carbon;

/**
 * The daily "we have started" tick for shaping and for baking.
 *
 * Each activity has a deadline; starting after it is flagged as late and
 * carries a salary-deduction warning. Whether a tick was late is decided
 * once, when it is recorded, and stored — changing the deadline setting
 * later must not rewrite history.
 */
class WorkStart extends Model
{
    use BelongsToBakery;

    public const CHANE = 'chane';

    public const BAKING = 'baking';

    public const TYPES = [
        self::CHANE => 'شروع چانه‌گیری',
        self::BAKING => 'شروع پخت',
    ];

    /**
     * Each activity is ticked by the one person who actually does it —
     * the chane gir starts shaping, the seller starts baking. Holding
     * record-work-start is not enough on its own.
     */
    public const RECORDED_BY = [
        self::CHANE => 'chane_gir',
        self::BAKING => 'seller',
    ];

    /** Used when the bakery has not set its own times. */
    public const DEFAULT_DEADLINES = [
        self::CHANE => '05:40',
        self::BAKING => '06:00',
    ];

    protected $fillable = [
        'type',
        'date',
        'started_at',
        'user_id',
        'is_late',
        'late_minutes',
        'late_sequence',
        'penalty_amount',
        'deadline',
        'note',
    ];

    protected function casts(): array
    {
        return [
            'date' => 'date',
            'started_at' => 'datetime',
            'is_late' => 'boolean',
            'late_minutes' => 'integer',
            'late_sequence' => 'integer',
            'penalty_amount' => 'decimal:2',
        ];
    }

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function scopeLate($query)
    {
        return $query->where('is_late', true);
    }

    public function scopeOfType($query, string $type)
    {
        return $query->where('type', $type);
    }

    public function getTypeLabelAttribute(): string
    {
        return self::TYPES[$this->type] ?? $this->type;
    }

    public function getStartedAtTimeAttribute(): ?string
    {
        return $this->started_at?->timezone(config('app.timezone'))->format('H:i');
    }

    public function getDateDisplayAttribute(): ?string
    {
        return AppCalendar::date($this->date);
    }

    /** The deadline this activity must start by, from the bakery settings. */
    public static function deadlineFor(string $type): string
    {
        $bakery = CurrentBakery::get();

        $column = $type === self::BAKING
            ? 'baking_start_deadline'
            : 'chane_start_deadline';

        $value = $bakery?->{$column};

        if (empty($value)) {
            return self::DEFAULT_DEADLINES[$type] ?? '06:00';
        }

        // The column is a TIME, which comes back as "05:40:00".
        return substr((string) $value, 0, 5);
    }

    /**
     * Records the tick for today, deciding lateness against the deadline.
     * Returns the existing record if the activity was already started.
     */
    public static function record(string $type, int $userId, ?Carbon $at = null): self
    {
        $now = ($at ?? now())->timezone(config('app.timezone'));
        $date = $now->toDateString();

        $existing = static::where('type', $type)->whereDate('date', $date)->first();

        if ($existing) {
            return $existing;
        }

        $deadline = self::deadlineFor($type);
        [$hour, $minute] = array_map('intval', explode(':', $deadline));

        $cutoff = $now->copy()->setTime($hour, $minute, 0);

        $late = $now->greaterThan($cutoff);

        // The tariff is charged per late day, not per late tick. If this day
        // has already been counted — because the other activity was late too
        // — it must not be charged a second time.
        $sequence = 0;
        $penalty = 0.0;

        if ($late) {
            $sequence = self::lateSequenceFor($now, $date);
            $penalty = self::alreadyCountedLate($date)
                ? 0.0
                : LatePenalty::amountFor($sequence);
        }

        return static::create([
            'type' => $type,
            'date' => $date,
            'started_at' => $now,
            'user_id' => $userId,
            'is_late' => $late,
            'late_minutes' => $late ? $cutoff->diffInMinutes($now) : 0,
            'late_sequence' => $sequence,
            'penalty_amount' => $penalty,
            'deadline' => $deadline,
        ]);
    }

    /** Whether this date was already recorded as late by another activity. */
    private static function alreadyCountedLate(string $date): bool
    {
        return static::whereDate('date', $date)->where('is_late', true)->exists();
    }

    /**
     * Which late day of the Jalali month this is — 1 for the first, and so
     * on. A day already marked late keeps its own number rather than taking
     * a new one.
     */
    public static function lateSequenceFor(Carbon $now, string $date): int
    {
        [$from, $to] = Jalali::monthRangeFor($now);

        $lateDates = static::where('is_late', true)
            ->whereBetween('date', [$from->toDateString(), $to->toDateString()])
            ->pluck('date')
            ->map(fn ($d) => $d instanceof Carbon ? $d->toDateString() : (string) $d)
            ->unique();

        // Same day already counted: reuse its position.
        if ($lateDates->contains($date)) {
            return $lateDates->sort()->values()->search($date) + 1;
        }

        return $lateDates->count() + 1;
    }

    /** This Jalali month's late days and what they have cost so far. */
    public static function monthSummary(?Carbon $at = null): array
    {
        $now = ($at ?? now())->timezone(config('app.timezone'));
        [$from, $to] = Jalali::monthRangeFor($now);

        $late = static::where('is_late', true)
            ->whereBetween('date', [$from->toDateString(), $to->toDateString()])
            ->get();

        $lateDays = $late->pluck('date')
            ->map(fn ($d) => $d instanceof Carbon ? $d->toDateString() : (string) $d)
            ->unique()
            ->count();

        $penalty = (float) $late->sum('penalty_amount');
        $free = LatePenalty::freeDays();

        return [
            'period_label' => AppCalendar::monthLabel($now),
            'late_days' => $lateDays,
            'free_days' => $free,
            'warnings_remaining' => max(0, $free - $lateDays),
            'penalty_total' => Money::convert($penalty),
            'penalty_total_formatted' => Money::format($penalty),
            // What the next late day would cost, so it is known beforehand.
            'next_day_amount_formatted' => Money::format(
                LatePenalty::amountFor($lateDays + 1)
            ),
        ];
    }

    /** The warning shown to staff when a tick was late. */
    public function getWarningAttribute(): ?string
    {
        if (! $this->is_late) {
            return null;
        }

        $message = 'اخطار: '.$this->type_label.' با '.$this->late_minutes
            .' دقیقه تأخیر نسبت به ساعت '.substr((string) $this->deadline, 0, 5)
            .' ثبت شد.';

        if ((float) $this->penalty_amount > 0) {
            return $message.' '.Money::format($this->penalty_amount).' کسر حقوق دارد'
                .' ('.$this->late_sequence.'اُمین تأخیر این ماه).';
        }

        if ($this->late_sequence > 0) {
            return $message.' '.LatePenalty::describe($this->late_sequence);
        }

        return $message;
    }

    /**
     * Today's board for both activities: started or not, late or not, and
     * how long is left before the deadline.
     *
     * On a day the shop is closed there is nothing to be late for, so no
     * deadline is reported at all.
     */
    public static function todayBoard(): array
    {
        $now = now()->timezone(config('app.timezone'));
        $isHoliday = Holiday::whereDate('date', $now->toDateString())->exists();

        $records = static::whereDate('date', $now->toDateString())
            ->with('user:id,name')
            ->get()
            ->keyBy('type');

        $items = [];

        foreach (self::TYPES as $type => $label) {
            $record = $records->get($type);
            $deadline = self::deadlineFor($type);
            [$hour, $minute] = array_map('intval', explode(':', $deadline));
            $cutoff = $now->copy()->setTime($hour, $minute, 0);

            $items[] = [
                'type' => $type,
                'label' => $label,
                'deadline' => $deadline,
                'started' => $record !== null,
                'started_at' => $record?->started_at_time,
                'started_by' => $record?->user?->name,
                'is_late' => (bool) $record?->is_late,
                'late_minutes' => (int) ($record?->late_minutes ?? 0),
                'warning' => $record?->warning,
                'is_holiday' => $isHoliday,
                // Negative once the deadline has passed without a tick.
                'minutes_remaining' => $record || $isHoliday
                    ? null
                    : (int) $now->diffInMinutes($cutoff, false),
                // Nothing recorded and the deadline gone: already late.
                'overdue' => ! $record && ! $isHoliday && $now->greaterThan($cutoff),
            ];
        }

        return [
            'date' => $now->toDateString(),
            'date_display' => AppCalendar::date($now),
            'is_holiday' => $isHoliday,
            'items' => $items,
            // The tariff is shown to every member of staff, not just to
            // whoever was late, so the rules are known in advance.
            'tariff' => LatePenalty::tariff(),
            'month_summary' => self::monthSummary($now),
        ];
    }
}
